src/ucp/rndv/rndv_rkey_ptr.c

```diff
@@ -183,3 +190,266 @@ ucp_proto_t ucp_rndv_rkey_ptr_proto = {
+
+    ucs_for_each_bit(md_index, md_map) {
+        md_attr = &context->tl_mds[md_index].attr;
+        if ((md_attr->cap.flags & UCT_MD_FLAG_RKEY_PTR) &&
+            /* Do not use xpmem, because cuda_copy registration will fail and
+             * performance will not be optimal. */
+            !(md_attr->cap.flags & UCT_MD_FLAG_REG) &&
+            (md_attr->cap.access_mem_types & UCS_BIT(remote_mem_type))) {
+            rkey_ptr_md_index = md_index;
+            break;
+        }
+    }
+
+    if (rkey_ptr_md_index == UCP_NULL_RESOURCE) {
+        ucp_proto_request_abort(req, UCP_NULL_RESOURCE);
+        return UCS_OK;
+    }
+
+    ppln_data = ucp_ep_peer_mem_get(context, req->send.ep, remote_address,
+                                    req->send.state.dt_iter.length, rkey_buffer,
+                                    rkey_ptr_md_index);
+    if (ppln_data->rkey == NULL) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    rkey_index = ucs_bitmap2idx(ppln_data->rkey->md_map, rkey_ptr_md_index);
+    status     = uct_rkey_ptr(ppln_data->rkey->tl_rkey[rkey_index].cmpt,
+                              &ppln_data->rkey->tl_rkey[rkey_index].rkey,
+                              req->send.rndv.remote_address, &local_ptr);
+    if (status != UCS_OK) {
+        ppln_data->size = 0; /* Make sure hash element is updated next time */
+        ucp_proto_request_abort(req, status);
+        return UCS_OK;
+    }
+
+    if (ppln_data->uct_memh == NULL) {
+        /* Register remote memory segment with memtype ep MD. Without
+        * registration fetching data from GPU to CPU will be performance
+        * inefficient. */
+        md_map              = 0;
+        ppln_data->md_index = ucp_ep_md_index(mem_type_ep, mem_type_rma_lane);
+
+        status = ucp_mem_rereg_mds(context, UCS_BIT(ppln_data->md_index),
+                                   local_ptr, ppln_data->size,
+                                   UCT_MD_MEM_ACCESS_RMA |
+                                           UCT_MD_MEM_FLAG_HIDE_ERRORS,
+                                   NULL, remote_mem_type, NULL,
+                                   &ppln_data->uct_memh, &md_map);
+
+        if (status != UCS_OK) {
+            ppln_data->md_index = UCP_NULL_RESOURCE;
+        } else {
+            ucs_assertv(md_map == UCS_BIT(ppln_data->md_index),
+                        "mdmap=0x%lx, md_index=%u", md_map,
+                        ppln_data->md_index);
+        }
+    }
+
+    ucp_proto_completion_init(&req->send.state.uct_comp,
+                              ucp_proto_rndv_rkey_ptr_mtype_copy_completion);
+
+    req->send.rndv.put.atp_map = rpriv->atp_map;
+
+    /* Set up IOV pointing to attached remote buffer */
+    ucs_assert(req->send.state.dt_iter.dt_class == UCP_DATATYPE_CONTIG);
+
+    iov.length = req->send.state.dt_iter.length;
+    iov.buffer = local_ptr;
+    iov.memh   = ppln_data->uct_memh;
+    iov.stride = 0;
+    iov.count  = 1;
+
+    status = uct_ep_get_zcopy(
```

Comment:
i think it should go thru `ppln` protocols (`ucp_proto_rndv_ppln_init`) when get_zcopy should be selected for fetching data from GPU to the remote shared buffer